### PR TITLE
[Data] Adding in default behavior to false for creating dirs on s3 writes

### PR DESCRIPTION
**Pull request by omatthew98**
## Why are these changes needed?
We should skip creating directories in s3 unless the user specifically overrides this behavior. PyArrow's s3fs implementation might collide with restrictive IAM permissions and skipping the create_dir call will prevent some of these errors.  

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by c21 — APPROVED** [review]

File: python/ray/data/context.py [hunk 1]
```diff
@@ -216,6 +220,8 @@ class DataContext:
             always written to the Ray Data log file.
         print_on_execution_start: If ``True``, print execution information when
             execution starts.
+        s3_try_create_dir: If ``True``, try to create directories on S3 when a write
+            call is made with a S3 URI.
```
Comment (bveeramani) [inline python/ray/data/context.py:224]: Should this go in `DataContext` directly or `ExecutionOptions` (not suggesting one way or another, just genuinely not sure about how we decide)

File: python/ray/data/datasource/file_datasink.py [hunk 2]
```diff
@@ -94,7 +95,17 @@ def on_write_start(self) -> None:
         """
         from pyarrow.fs import FileType
 
-        if self.try_create_dir:
+        # We should skip creating directories in s3 unless the user specifically
+        # overrides this behavior. PyArrow's s3fs implementation might collide
+        # with restrictive IAM permissions and skipping the create_dir call will
```
Comment (bveeramani) [inline python/ray/data/datasource/file_datasink.py:100]: > "PyArrow's s3fs implementation might collide with restrictive IAM permissions"

Could you elaborate on this in the source code? Don't think I would've understood this without our offline conversation

File: python/ray/data/datasource/file_datasink.py [hunk 3]
```diff
@@ -94,7 +95,17 @@ def on_write_start(self) -> None:
...
+        # prevent some of these errors.
+        parsed_uri = urlparse(self.path)
+        is_s3_uri = parsed_uri.scheme == "s3"
```
Comment (bveeramani) [inline python/ray/data/datasource/file_datasink.py:103]: I think we can also do `isinstance(self.filesystem, S3Filsystem)`?

**Review by bveeramani — APPROVED** [review]

File: python/ray/data/context.py (hunk 1 above)
Comment (omatthew98) [reply python/ray/data/context.py:224]: Not sure if I have the context to know which would be preferable for this, will leave unresolved if anyone has strong feelings please weight in.

File: python/ray/data/datasource/file_datasink.py (hunk 2 above)
Comment (omatthew98) [reply python/ray/data/datasource/file_datasink.py:100]: Yeah, will update to make this more specific.

File: python/ray/data/datasource/file_datasink.py (hunk 3 above)
Comment (omatthew98) [reply python/ray/data/datasource/file_datasink.py:103]: I was following the logic [here](https://github.com/ray-project/ray/blob/fecc93994a5331378158d8bc5e6fe37d52436767/python/ray/_private/utils.py#L1819-L1822):
```python
parsed_uri = urlparse(uri)
    if parsed_uri.scheme == "s3":
        uri = _add_url_query_params(uri, {"allow_bucket_creation": True})
```

I think what you suggested should work for most cases but I think the above is slightly more general, primarily for cases where the user might not be using the S3FileSystem e.g. they are using a custom implementation that interacts with s3 objects.

**Review by c21 — APPROVED** [review]

**Pull request merged**